tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_connection_pool.hpp

```diff
@@ -0,0 +1,653 @@
+    : public RedisVirtualWrapper {
+private:
+  SentinelOptions sentinel_opts;
+  ConnectionOptions conn_opts;
+  ConnectionPoolOptions pool_opts;
+
+public:
+  std::shared_ptr<RedisInstance> redis_conn; // for the hungry singleton mode
+
+public:
+  RedisWrapper(RedisInstance &&) = delete;
+  RedisWrapper(const RedisInstance &) = delete;
+  RedisWrapper &operator=(const RedisInstance &) = delete;
+
+  RedisWrapper() // In singleton mode, classes should not be initialized
+                 // through constructor
+  {
+    std::cout << "RedisSentinel connection pool constructor called!"
+              << std::endl;
+  }
+
+  ~RedisWrapper() {
+    if (redis_conn == nullptr) {
+      return;
+    }
+    redis_conn.reset();
+    std::cout << "RedisSentinel connection pool destructor called!"
+              << std::endl;
+  }
+
+public:
+  std::shared_ptr<RedisInstance> start_conn() {
+    sentinel_opts.nodes = {{redis_connection_params.redis_host_ip,
+                            redis_connection_params.redis_host_port}};
+    // Optional. Timeout before we successfully connect to Redis Sentinel.
+    sentinel_opts.connect_timeout = std::chrono::milliseconds(
+        redis_connection_params.redis_sentinel_connect_timeout);
+    // Optional. Timeout before we successfully send request to or receive
+    // response from Redis Sentinel.
+    sentinel_opts.socket_timeout = std::chrono::milliseconds(
+        redis_connection_params.sentinel_socket_timeout);
+
+    // Redis connection options
+    conn_opts.password =
+        redis_connection_params
+            .redis_password; // Optional. No redis_password by default.
+    conn_opts.db = redis_connection_params.redis_db;
+    conn_opts.connect_timeout = std::chrono::milliseconds(
+        redis_connection_params.redis_connect_timeout);
+    conn_opts.socket_timeout =
+        std::chrono::milliseconds(redis_connection_params.redis_socket_timeout);
+    // Redis connection pool options
+    pool_opts.size = redis_connection_params.redis_conn_pool_size;
+    pool_opts.wait_timeout =
+        std::chrono::milliseconds(redis_connection_params.redis_wait_timeout);
+    pool_opts.connection_lifetime =
+        std::chrono::minutes(redis_connection_params.redis_connection_lifetime);
+
+    auto sentinel = std::make_shared<Sentinel>(sentinel_opts);
+
+    try {
+      static auto redis_client = std::make_shared<RedisInstance>(
+          RedisInstance(sentinel, redis_connection_params.redis_master_name,
+                        Role::MASTER, conn_opts, pool_opts));
+      redis_client->ping();
+      return redis_client;
+    } catch (const std::exception &err) {
+      std::cerr << "RedisHandler--other " << err.what() << std::endl;
+      return nullptr;
+    } catch (...) {
+      std::cerr << "RedisHandler--other crash" << std::endl;
+      return nullptr;
+    }
+    return nullptr;
+  }
+
+  virtual void conn() override {
+    if (isRedisConnect == false) {
+      for (short i = 0; i < 10; i++) {
```

Comment:
10 ？
Is it better to be configurable?

Reply:
This relink is only used to establish a link to Redis for the first time, with more or less no impact on the business logic. I actually wanted to set it to 3.